src/core/exec-invoke.c

```diff
@@ -2440,6 +2448,91 @@ static int create_many_symlinks(const char *root, const char *source, char **sym
         return 0;
 }
 
+static int apply_exec_quotas(const char *target_dir, const char *cgroup_path, ExecDirectoryType type, uint32_t percent_quota) {
+        assert(target_dir);
+        assert(percent_quota > 0);
+        int r = 0;
+
+        dev_t devno;
+        r = get_block_device(target_dir, &devno);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to determine block device backing %s: %m", target_dir);
+        else if (devno == 0) {
+                log_warning("Not applying storage quotas for %s: File system not backed by a block device.", target_dir);
+                return 0;
+        }
+
+        /* Generate projid: cgroup path + exec directory type */
+        size_t buffer_size = strlen(cgroup_path) + snprintf(NULL, 0, "%d", INT_MIN) + 1;
+        char proj_id_plain[buffer_size];
+
+        r = snprintf(proj_id_plain, buffer_size, "%s%d", cgroup_path, type);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to generate project id for %s: %m", target_dir);
+        else if ((size_t) r >= buffer_size) {
+                log_debug("Exec quotas: Failed to generate project id for %s", target_dir);
+                return -1;
+        }
+
+        uint32_t proj_id = MurmurHash2(proj_id_plain, strlen(proj_id_plain), 0);
+
+        /* Check if project quotas are supported */
+        struct dqblk req;
+        r = quotactl_devnum(QCMD_FIXED(Q_GETQUOTA, PRJQUOTA), devno, proj_id, &req);
+        if (r == -ESRCH || ERRNO_IS_NEG_NOT_SUPPORTED(r)) {
+                log_warning("Not applying storage quotas: project quotas are not supported for %s", target_dir);
+                return 0;
+        } else if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to query disk quota for prjid %u: %m", proj_id);
+
+        /* Set projid (equivalent to chattr +P -p) */
+        struct fsxattr attrs;
+        int fd = xopenat(AT_FDCWD, target_dir, O_RDONLY|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW);
+        r = ioctl(fd, FS_IOC_FSGETXATTR, &attrs);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to get fsxattr for %s", target_dir);
+
+        attrs.fsx_xflags = FS_XFLAG_PROJINHERIT;
+        attrs.fsx_projid = proj_id;
+
+        r = ioctl(fd, FS_IOC_FSSETXATTR, &attrs);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to set project id for %s: %m", target_dir);
```

Comment:
could you please add a nice helper for setting the projid, maybe to src/basic/chattr.[ch] or src/shared/quota-util.[ch] or so